Document Module - Render Table Of Contents on server-side

ahmetcelik05 opened the pull request:
### Description

Resolves #23643 

### Checklist

- [ ] I fully tested it as developer
- [ ] no need to document

### How to test it?

This PR should be tested in conjunction with the corresponding [vs-internal PR](https://github.com/volosoft/vs-internal/pull/7280).

ahmetcelik05 committed:
Implement server-side Table of Contents generation

Replaces client-side TOC generation with a new ITocGeneratorService and TocGeneratorService using HtmlAgilityPack. Updates the project page to render the TOC from the server, removes bootstrap-toc dependencies, and adjusts related JS and CSS for the new TOC structure. Adds HtmlAgilityPack as a dependency.

ahmetcelik05 committed:
Merge branch 'dev' into Issue-#23643

ahmetcelik05 committed:
Refactor TOC heading handling with Heading record

Introduces a Heading record to replace tuple usage for TOC headings, improving code readability and maintainability. Updates related logic in TocGeneratorService to use the new Heading type and adds constants for heading levels.

ahmetcelik05 committed:
Refactor TOC generation to use Markdig and remove HtmlAgilityPack

Replaces the previous HTML-based table of contents (TOC) extraction using HtmlAgilityPack with a Markdig-based approach. Introduces custom Markdig extensions and renderers to extract headings directly from markdown, updates the TOC service and interface, and removes the HtmlAgilityPack dependency from the project.

ahmetcelik05 committed:
Update TocGeneratorService.cs

ahmetcelik05 committed:
Update TocGeneratorService.cs

ahmetcelik05 committed:
Refactor TOC generation to use heading data model

Replaces HTML-based TOC generation with a model-based approach by extracting heading data from markdown and rendering the table of contents via a new partial view. Removes the custom Markdig extension and related classes, updates the service interface, and adapts the page model and views to use the new heading list.

ahmetcelik05 committed:
Refactor TOC heading to use TocHeading record

Replaces the internal Heading record in TocGeneratorService with a new public TocHeading record class. Updates all references and method signatures to use TocHeading, improving code clarity and reusability.

salihozkara commented on modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/Index.cshtml:
```diff
@@ -597,8 +597,8 @@
-                            <nav id="docs-sticky-index" class="navbar index-scroll">
-                                @Html.Raw(Model.TocHtml)
+                            <nav id="toc" class="navbar index-scroll">
+                                    <partial name="TableOfContents" model="Model" />
```
```suggestion
                                    <partial name="TableOfContents" model="Model.TocHeadings" />
```

salihozkara commented on modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/TableOfContents.cshtml:
```diff
@@ -0,0 +1,73 @@
+﻿@model Volo.Docs.Pages.Documents.Project.IndexModel
```
```suggestion
﻿@model List<Heading>

```

salihozkara commented on modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/TableOfContents.cshtml:
```diff
@@ -0,0 +1,73 @@
+﻿@model Volo.Docs.Pages.Documents.Project.IndexModel
+@{
+    if (Model.TocHeadings == null ||Model.TocHeadings.Count == 0)
```
```suggestion
    if (Model == null || Model.Count == 0)

```

salihozkara commented on modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/TableOfContents.cshtml:
```diff
@@ -0,0 +1,73 @@
+        return;
+    }
+
+    var relevantHeadings = Model.TocHeadings
```
```suggestion
    var relevantHeadings = Model

```

salihozkara commented on modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/TableOfContents.cshtml:
```diff
@@ -0,0 +1,73 @@
+
+    if (relevantHeadings.Count == 0)
+    {
+        relevantHeadings = Model.TocHeadings
```
```suggestion
        relevantHeadings = Model

```

salihozkara commented on modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/TableOfContents.cshtml.cs:
delete this file

ahmetcelik05 committed:
Refactor TableOfContents partial to use TocHeadings model

Updated the TableOfContents partial to accept a List<TocHeading> instead of the full IndexModel, simplifying its interface. Adjusted the Index.cshtml to pass Model.TocHeadings directly. Removed the unused TableOfContents.cshtml.cs code-behind file.

ahmetcelik05 committed:
Merge branch 'dev' into Issue-#23643

salihozkara committed:
Refactor table of contents to hierarchical structure

Replaces flat TocHeading list with hierarchical TocItem structure for document table of contents. Updates rendering logic, service interface, and implementation to support nested headings and configurable levels.

salihozkara committed:
Refactor and improve TOC generation logic

Enhanced TocGeneratorService by refactoring heading parsing, optimizing inline text extraction, and improving handling of empty or invalid markdown. Added helper methods for pipeline creation, heading construction, and inline processing for better readability and maintainability.

ahmetcelik05 committed:
Update Index.cshtml.cs

ahmetcelik05 committed:
Refactor TOC generation to use level count instead of max level

Replaces the use of a fixed max TOC level with a level count parameter in TOC generation methods and logic. This change improves flexibility by allowing the number of heading levels included in the TOC to be specified, rather than a maximum heading level value.
